### DISCUSSION TITLE
Why is dispatchWorkgroupsIndirect() running much slower than dispatchWorkgroups()?

### DISCUSSION THREAD

**Discussion by Loner789** [gpuweb/gpuweb#4463] [Q&A]
Hi everybody! I am programming a moving heuristic algorithm on webgpu, and therefore I create a loop inside the encoder that initializes 4 consecutive shaders n times until finding the appropriate position for all the elements that need to be placed. I have the following question: why can dispatchWorkgroupsIndirect() with an individual value for every element run much slower (10 times slower) than dispatchWorkgroups() with a hardcoded value for all elements? As far as I understand, the opposite should happen. Can someone tell me why it works this way?

**Comment by conorpo** [comment]
@toji actually just wrote an article about this: https://toji.dev/webgpu-best-practices/indirect-draws

**Comment by toji** [comment]
Actually, the validation in that article is specifically related to indirect draws, not compute dispatch. I'm not sure what would account for the overhead here. @Loner789, do you happen to have a test page we could look at that demonstrates the problem?

(Also, assuming that it ends up being an implementation issue we'll probably ask that this gets filed as an issue on the appropriate browser's bug tracker. This repo's issue tracker is for issues with the spec.)
